apps/mobile/src/features/send/components/recipient/v2/recipient-evaluator.ts

```diff
@@ -0,0 +1,130 @@
+import { recipientSchemaResultContainsError } from '@/features/send/components/recipient/v2/recipient.utils';
+import { t } from '@lingui/macro';
+import { ZodSchema } from 'zod';
+
+interface CreateRecipientEvaluatorParams {
+  schema: ZodSchema;
+  isNewAddress: (address: string) => boolean;
+  canSelfSend: boolean;
+}
+
+type GuardReason =
+  | 'newRecipient'
+  | 'recipientIsPayer'
+  | 'incorrectNetworkAddress'
+  | 'nonCompliantAddress';
+
+interface GuardResultOk {
+  severity: 'none';
+}
+
+interface GuardResultWarnOrBlock {
+  severity: 'warn' | 'block';
+  address: string;
+  reason: GuardReason;
+  title: string;
+  description: string;
+  primaryActionLabel?: string;
+  secondaryActionLabel?: string;
+}
+
+export type GuardResult = GuardResultOk | GuardResultWarnOrBlock;
+
+export function createRecipientEvaluator({
+  canSelfSend,
+  isNewAddress,
+  schema,
+}: CreateRecipientEvaluatorParams) {
+  return async function evaluateRecipient(address: string): Promise<GuardResult> {
+    const validationResult = await schema.safeParseAsync(address);
+    const newRecipient = isNewAddress(address);
+
+    if (recipientSchemaResultContainsError(validationResult, 'recipientIsPayer')) {
+      if (canSelfSend) {
```

Comment:
Nice work here 👍 I was going to suggest doing this another way but on review of the other code this is the best way. 

We could call it something like `canConsolidate` or something but `canSelfSend` is fine I think

Reply:
The best way would probably be not using `zod` altogether. The way this flow turned out made both `zod` and `react-hook-form` mostly redundant--those mostly just add friction now with almost no value. Something to look into way down the line.

> canConsolidate
I thought about this as well but consolidation is a single a special case, where as self send is just a likely mistake we block.